app/src/main/java/org/thoughtcrime/securesms/util/ProfileUtil.java

```diff
@@ -0,0 +1,333 @@
+  }
+
+  @WorkerThread
+  public static @NonNull MobileCoinPublicAddress getAddressForRecipient(@NonNull Recipient recipient)
+      throws IOException, PaymentsAddressException
+  {
+    ProfileKey profileKey;
+    try {
+      profileKey = getProfileKey(recipient);
+    } catch (IOException e) {
+      Log.w(TAG, "Profile key not available for " + recipient.getId());
+      throw new PaymentsAddressException(PaymentsAddressException.Code.NO_PROFILE_KEY);
+    }
+    ProfileAndCredential profileAndCredential     = ProfileUtil.retrieveProfileSync(ApplicationDependencies.getApplication(), recipient, SignalServiceProfile.RequestType.PROFILE);
+    SignalServiceProfile profile                  = profileAndCredential.getProfile();
+    byte[]               encryptedPaymentsAddress = profile.getPaymentAddress();
+
+    if (encryptedPaymentsAddress == null) {
+      Log.w(TAG, "Payments not enabled for " + recipient.getId());
+      throw new PaymentsAddressException(PaymentsAddressException.Code.NOT_ENABLED);
+    }
+
+    try {
+      IdentityKey                        identityKey             = new IdentityKey(Base64.decode(profileAndCredential.getProfile().getIdentityKey()), 0);
+      ProfileCipher                      profileCipher           = new ProfileCipher(profileKey);
+      byte[]                             decrypted               = profileCipher.decryptWithLength(encryptedPaymentsAddress);
+      SignalServiceProtos.PaymentAddress paymentAddress          = SignalServiceProtos.PaymentAddress.parseFrom(decrypted);
+      byte[]                             bytes                   = MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(paymentAddress, identityKey);
+      MobileCoinPublicAddress            mobileCoinPublicAddress = MobileCoinPublicAddress.fromBytes(bytes);
+
+      if (mobileCoinPublicAddress == null) {
+        throw new PaymentsAddressException(PaymentsAddressException.Code.INVALID_ADDRESS);
+      }
+
+      return mobileCoinPublicAddress;
+    } catch (InvalidCiphertextException | IOException e) {
+      Log.w(TAG, "Could not decrypt payments address, ProfileKey may be outdated for " + recipient.getId(), e);
+      throw new PaymentsAddressException(PaymentsAddressException.Code.COULD_NOT_DECRYPT);
+    } catch (InvalidKeyException e) {
+      Log.w(TAG, "Could not verify payments address due to bad identity key " + recipient.getId(), e);
+      throw new PaymentsAddressException(PaymentsAddressException.Code.INVALID_ADDRESS_SIGNATURE);
+    }
+  }
+
+  private static ProfileKey getProfileKey(@NonNull Recipient recipient) throws IOException {
+    byte[] profileKeyBytes = recipient.getProfileKey();
+
+    if (profileKeyBytes == null) {
+      Log.w(TAG, "Profile key unknown for " + recipient.getId());
+      throw new IOException("No profile key");
+    }
+
+    ProfileKey profileKey;
+    try {
+      profileKey = new ProfileKey(profileKeyBytes);
+    } catch (InvalidInputException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)